src/mame/hp/hp98x6.cpp

```diff
@@ -696,6 +807,1236 @@ ROM_START(hp9816a)
+
+	// ~225 ms duration
+	TTL74123(mconfig, m_ss[0], RES_K(23.7), CAP_U(33));
+	m_ss[0]->set_connection_type(TTL74123_NOT_GROUNDED_NO_DIODE);
+	m_ss[0]->set_a_pin_value(0);
+	m_ss[0]->set_clear_pin_value(1);
+	TTL74123(mconfig, m_ss[1], RES_K(23.7), CAP_U(33));
+	m_ss[1]->set_connection_type(TTL74123_NOT_GROUNDED_NO_DIODE);
+	m_ss[1]->set_a_pin_value(0);
+	m_ss[1]->set_clear_pin_value(1);
+
+	TIMER(mconfig, m_fdc_timer).configure_generic(FUNC(hp9826_36_state::fdc_ram_io));
+}
+
+void hp9826_36_state::cpu_mem_map(address_map &map)
+{
+	hp98x6_base_state::cpu_mem_map(map);
+
+	// Range           DTACK Device
+	// ==============================
+	// 40'0000-41'ffff *     None
+	// 42'0000-42'ffff *     Keyboard UPI
+	// 43'0000-43'ffff *     None
+	// 44'0000-44'ffff *     Floppy
+	// 45'0000-46'ffff *     None
+	// 47'0000-47'ffff *     HPIB
+	// 48'0000-4f'ffff *     None
+	// 50'0000-5e'ffff       None
+	// 5f'0000-5f'3fff *     ID PROM & DIP switches
+	// 5f'4000-fb'ffff       None
+	map(0x400000, 0x41ffff).noprw();
+	map(0x420000, 0x42ffff).mask(3).rw(m_upi, FUNC(hp98x6_upi_device::read), FUNC(hp98x6_upi_device::write)).umask16(0x00ff);
+	map(0x430000, 0x43ffff).noprw();
+	map(0x440000, 0x44ffff).rw(FUNC(hp9826_36_state::floppy_r), FUNC(hp9826_36_state::floppy_w)).umask16(0xff00);
+	map(0x450000, 0x46ffff).noprw();
+	map(0x470000, 0x47ffff).rw(FUNC(hp9826_36_state::hpib_r), FUNC(hp9826_36_state::hpib_w)).umask16(0x00ff);
+	map(0x480000, 0x4fffff).noprw();
+	map(0x500000, 0x5effff).rw(m_cpu, FUNC(m68000_device::berr_r), FUNC(m68000_device::berr_w));
+	map(0x5f0000, 0x5f3fff).nopw().r(FUNC(hp9826_36_state::id_prom_r));
+	map(0x5f4000, 0xfbffff).rw(m_cpu, FUNC(m68000_device::berr_r), FUNC(m68000_device::berr_w));
+}
+
+void hp9826_36_state::cpu_reset_w(int state)
+{
+	hp98x6_base_state::cpu_reset_w(state);
+
+	if (state) {
+		floppy_reset();
+	}
+}
+
+unsigned hp9826_36_state::get_sel_floppy() const
+{
+	return BIT(m_floppy_ctrl, FC_DRIVE_SEL_BIT) ? 1 : 0;
+}
+
+floppy_image_device *hp9826_36_state::get_drive(unsigned idx) const
+{
+	if (idx) {
+		return bool(m_drive1) ? m_drive1->get_device() : nullptr;
+	} else {
+		return m_drive0->get_device();
+	}
+}
+
+unsigned hp9826_36_state::get_floppy_idx(floppy_image_device *floppy) const
+{
+	return get_drive(0) == floppy ? 0 : 1;
+}
+
+void hp9826_36_state::floppy_reset()
+{
+	m_floppy_ctrl = 0;
+	m_fdc->mr_w(0);
+	floppy_update_sel();
+}
+
+void hp9826_36_state::floppy_update_sel()
+{
+	floppy_image_device *drives[2] = { get_drive(0), get_drive(1) };
+	floppy_image_device *new_floppy = drives[get_sel_floppy()];
+
+	if (new_floppy != m_curr_floppy) {
+		m_curr_floppy = new_floppy;
+		m_fdc->set_floppy(m_curr_floppy);
+	}
+	if (m_curr_floppy) {
+		m_curr_floppy->setup_index_pulse_cb(floppy_image_device::index_pulse_cb(&hp9826_36_state::floppy_index_cb, this));
+		m_curr_floppy->ss_w(BIT(m_floppy_ctrl, FC_SIDE_SEL_BIT));
+	}
+	floppy_update_mon_hlt();
+}
```

Comment:
It looks like you’re going to need to implement `device_post_load()` to correct `m_curr_floppy` after loading a saved state, or it could get out of sync with the rest of the state.